security: Initial gssapi support

BenPope opened the pull request:
Initial end-to-end test and feature for GSSAPI support.

This contains lots of debug code, and is underdocumented, but essentially it works:

TODO:
* [x] Play more nicely with seastar - krb5 now runs in the `worker_thread`
* [ ] Configuration for `krb5.conf` - I think this can wait.
* [ ] More wrappers for gssapi types - some failures still result in leaks
* [x] Logging cleanup (some trace logging results in binary written to the log)
* [ ] Much more testing

## Backports Required

- [x] none - not a bug fix
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v22.3.x
- [ ] v22.2.x
- [ ] v22.1.x

## UX Changes

GSSAPI is now a supported SASL mechanism on the Kafka API.

To enable the feature:
```sh
curl -sv -X PUT http://127.0.0.1:9644/v1/features/kafka_gssapi -d '{"state":"active"}'
```
To disable the feature:
```sh
curl -sv -X PUT http://127.0.0.1:9644/v1/features/kafka_gssapi -d '{"state":"disabled"}'
```
To enable GSSAPI as a support SASL mechanism:
```sh
rpk cluster config set sasl_mechanisms '["SCRAM","GSSAPI"]' --api-urls 127.0.0.1:9644
```

## Release Notes

### Features

* Support GSSAPI as a SASL mechanism on the Kafka API.

dotnwat commented on src/v/kafka/server/handlers/sasl_handshake.cc:
```diff
@@ -32,23 +62,22 @@ ss::future<response_ptr> sasl_handshake_handler::handle(
      * authenticate request for the next phase of the auth process.
      */
     auto error = error_code::none;
-
-    if (request.data.mechanism == security::scram_sha256_authenticator::name) {
-        ctx.sasl()->set_mechanism(
-          std::make_unique<security::scram_sha256_authenticator::auth>(
-            ctx.credentials()));
-
-    } else if (
-      request.data.mechanism == security::scram_sha512_authenticator::name) {
-        ctx.sasl()->set_mechanism(
-          std::make_unique<security::scram_sha512_authenticator::auth>(
-            ctx.credentials()));
-
+    auto mechs = supported_mechanisms();
+    auto const& it = mechs.find(request.data.mechanism);
+    if (it != mechs.end()) {
+        ctx.sasl()->set_mechanism(it->second(ctx.credentials()));
```
does this build instances of all the mechanisms but then only choose one of them? if so, why not a few if/else statements instead?

BenPope replied:
It builds a map of factories.

BenPope replied:
I've simplified this a bit.

dotnwat commented on src/v/security/CMakeLists.txt:
```diff
@@ -3,6 +3,7 @@ v_cc_library(
   SRCS
     scram_algorithm.cc
     scram_authenticator.cc
+    gssapi_authenticator.cc
```
i don't think this file has been added yet

BenPope replied:
Good spot, fixed.

dotnwat commented on src/v/security/gssapi.h:
```diff
@@ -0,0 +1,152 @@
+
+namespace security::gss {
+
+struct buffer_view : ::gss_buffer_desc {
```
kinda feel like we should encapsulate rather than inherit in cases like this where we'll be interacting with C libraries (e.g. gss_release_buffer). wdyt?

BenPope replied:
Done

dotnwat commented on src/v/security/gssapi.h:
```diff
@@ -0,0 +1,152 @@
+};
+
+struct ctx_id {
+    gss_ctx_id_t _impl{nullptr};
```
would it make sense to create the context immediately on construction?

dotnwat commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,511 @@
+    return ss.str();
+}
+
+struct hex {
+    explicit hex(bytes const& v)
+      : _v{v} {}
+    explicit hex(bytes_view const& v)
+      : _v{v} {}
+    friend std::ostream& operator<<(std::ostream& os, hex const& h) {
+        for (size_t i = 0; i < h._v.length(); ++i) {
+            if ((i % 16) == 0) {
+                os << '\n';
+            }
+            os << std::setfill('0') << std::setw(2) << std::hex
+               << (int)(h._v[i]) << ' ';
+        }
+        return os;
+    }
+    bytes_view _v;
+};
```
i was thinking we had some hex utilities already. should this go with those (if they exist?)

BenPope replied:
Thanks. Should probably one-line this if I keep it.

BenPope replied:
I've removed this and logged the length instead.

dotnwat commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,511 @@
+        _state = state::failed;
+        return errc::invalid_scram_state;
+    }
+    using namespace gss;
```
remove?

dotnwat commented on src/v/security/gssapi_authenticator.h:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    const security::acl_principal& principal() const override {
+        vlog(seclog.info, "gss principal: {} {}", _principal, _state);
```
curious to have the vlog in what seems like a simple accessor?

BenPope replied:
Removed

dotnwat reviewed (commented):
this is awesome. i think all my feedback is minor stuff.

BenPope commented on src/v/config/configuration.cc:
```diff
@@ -823,6 +823,25 @@ configuration::configuration()
+  , sasl_kerberos_principal(
+      *this,
+      "sasl_kerberos_principal",
+      "The principal to use",
```
Improved

michael-redpanda commented on src/v/security/gssapi.h:
```diff
@@ -0,0 +1,191 @@
+    buffer operator=(buffer&&) = delete;
+    buffer operator=(buffer const&) = delete;
+
+    ~buffer() {
+        if (value() != nullptr) {
+            OM_uint32 minor{};
+            gss_release_buffer(&minor, &(*this));
+        }
```
Should this destructor live with `buffer_view`?

michael-redpanda commented on src/v/security/gssapi.h:
```diff
@@ -0,0 +1,191 @@
+
+namespace security::gss {
+
+class buffer_view {
+public:
+    explicit buffer_view(::gss_buffer_desc bd)
+      : _impl{bd} {}
+    explicit buffer_view(std::string_view sv)
+      : _impl{sv.size(), const_cast<char*>(sv.data())} {}
+    explicit buffer_view(bytes_view bv)
+      : _impl{bv.size(), const_cast<unsigned char*>(bv.data())} {}
+
+    explicit operator std::string_view() const {
+        return {static_cast<char*>(_impl.value), _impl.length};
+    };
+    explicit operator bytes_view() const {
+        return {static_cast<unsigned char*>(_impl.value), _impl.length};
+    };
+
+    const void* value() const { return _impl.value; }
+    size_t size() const { return _impl.length; }
+    const gss_buffer_desc* operator&() const { return &_impl; };
+    // TODO BP: Should I wrap every call that only uses this as an input param?
+    gss_buffer_desc* operator&() { return &_impl; };
+
+private:
+    ::gss_buffer_desc _impl;
+};
```
Does it need to be explicit that `buffer_view` doesn't 'own' the `gss_buffer_desc` that it contains (unlike `buffer` below)?  What if there's a call to `gss_release_buffer` on a copy of the buffer that `buffer_view` contains?

BenPope replied:
The `buffer_view` is non-owning, like `std::string_view`, typically used as an adaptor for input parameters to GSSAPI calls. The type makes it explicit, did you have something else in mind?

The conversion operator can't be const because c. Wrapping all of the calls and `const_cast`ing is a fair bit of effort, they're mostly called just once.

michael-redpanda replied:
Oh ok the suffix `_view` implies non-ownership then?  I haven't worked much with `std::string_view` so maybe that's the piece I was missing.  Appreciate the explanation.

BenPope replied:
I guess. We also have `bytes_view` https://github.com/redpanda-data/redpanda/blob/dev/src/v/bytes/bytes.h#L33, which is why there are conversion operators to `std::string_view` and `bytes_view`. The standard has `std::span` for similar, but we don't really use that much, so I just went with `_view` suffix.

michael-redpanda commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+        OM_uint32 min_stat{};
+        gss::buffer msg;
+        auto maj_stat = gss_display_status(
+          &min_stat, code, type, GSS_C_NO_OID, &msg_ctx, &msg);
```
Do we care of msg_ctx != 0?  I believe, if I'm reading the docs correctly (https://docs.oracle.com/cd/E19455-01/816-3322/6m9k32d10/index.html) a non-0 return here indicates there may be more info to get back?

BenPope replied:
Valid question. Much of this is copy-pasta from examples online. It's my understanding that it's not relevant for the use case of the plugins we're using, but it's a touch under-documented.

michael-redpanda replied:
Yeah I can see the challenge you went through for this

BenPope replied:
I'd prefer to have an actually nice API and a `fmt::formatter` specialisation for a return type/exception/result that contained the major,minor pair.

dotnwat replied:
i think there is a decent example of this API being used in Ceph, fwiw.

michael-redpanda replied:
Per the man page:
> gss_display_status() will return a non-zero value in this parameter if there are further messages to extract

So yes, if `gss_display_status` sets `msg_ctx` to a non-0 number, there is more status information to extract.

michael-redpanda commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+    if (flags & GSS_C_REPLAY_FLAG) ss << "context flag: GSS_C_REPLAY_FLAG\n";
+    if (flags & GSS_C_SEQUENCE_FLAG)
+        ss << "context flag: GSS_C_SEQUENCE_FLAG\n";
+    if (flags & GSS_C_CONF_FLAG) ss << "context flag: GSS_C_CONF_FLAG \n";
+    if (flags & GSS_C_INTEG_FLAG) ss << "context flag: GSS_C_INTEG_FLAG \n";
```
_very minor_ nit: the other items don't have a space before the newline

BenPope replied:
Done

BenPope commented on src/v/config/configuration.h:
```diff
@@ -177,6 +177,9 @@ struct configuration final : public config_store {
     property<int16_t> id_allocator_log_capacity;
     property<int16_t> id_allocator_batch_size;
     property<bool> enable_sasl;
+    property<ss::sstring> sasl_mechanisms;
```
I've changed this on my branch after feedback

michael-redpanda commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+    if (major_status != GSS_S_COMPLETE) {
+        display_status("import name", major_status, minor_status);
+        _state = state::failed;
+        return errc::invalid_scram_state;
```
Probably a tbd but should these return codes be renamed (namely since this isn't scram that's causing the error)?

BenPope replied:
Oh, good, point. Noted.

BenPope replied:
Done

dotnwat commented on src/v/security/gssapi.h:
```diff
@@ -0,0 +1,191 @@
+    ~buffer() {
+        if (value() != nullptr) {
+            OM_uint32 minor{};
+            gss_release_buffer(&minor, &(*this));
```
nit:

this looks so error prone to me like `&(*this)` isn't what it should be. of course, because of the operator overload it is. but it makes it look like i should be able to `gss_release_buffer(&minor, this)`. anyway, wrapping c apis is always fun. maybe do this in the super class destructor and access impl directly?

dotnwat commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+
+std::string display_ctx_flags(OM_uint32 flags) {
+    std::stringstream ss;
+    if (flags & GSS_C_DELEG_FLAG) ss << "context flag: GSS_C_DELEG_FLAG\n";
```
cool. i didn't clang-format allowed single-line conditionals.

dotnwat commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+        auto principal = config::shard_local_cfg().sasl_kerberos_principal();
+        auto keytab = config::shard_local_cfg().sasl_kerberos_keytab();
+
+        if (auto res = co_await _worker.submit(
+              [this, principal, keytab]() { return init(principal, keytab); });
```
wild

dotnwat commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+    }
+        [[fallthrough]];
+    case state::more: {
+        auto res = co_await _worker.submit(
+          [this, auth_bytes]() { return more(auth_bytes); });
```
does the worker serialize all operations? i'm thinking about how `more` is a class method and might access shared state?

BenPope replied:
Yeah, they're all backed up on the semaphore.

dotnwat commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+    }
+    vlog(
+      seclog.trace,
+      "gss {} gss_export_cred: {}",
+      _state,
+      hex(bytes_view{token}));
```
i don't know exactly what all the state is that's being logged, but i recall we need to be careful about logging anything that might be considered sensitive.

BenPope replied:
Replaced by length

dotnwat commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+      seclog.trace,
+      "recv_tok: length: {} - {}",
+      auth_bytes.length(),
+      hex{auth_bytes});
```
like maybe raw bytes?

BenPope replied:
Replaced by length

dotnwat commented on tests/rptest/services/kerberos.py:
```diff
@@ -0,0 +1,222 @@
+KDC_DB_PATH = "/var/lib/krb5kdc/principal"
+
+
+class KrbKdc(Service):
```
:heart:

dotnwat reviewed (commented):
yeh just some minor stuff. it looks really good.

probably the only thing blocking this PR (some of the follow ups can be in separate PR) is making sure the worker thread is safe. added some comments about that in the review.

BenPope commented on src/v/config/configuration.cc:
```diff
@@ -823,6 +823,25 @@ configuration::configuration()
+  , sasl_kerberos_keytab(
+      *this,
+      "sasl_kerberos_keytab",
+      "The location of the Kerberos keytab file for Redpanda",
```
Yes. The principal (Service Principal Name) is made of three parts: `<primary>/<host>@<realm>`, the primary is configured with `sasl_kerberos_principal`.

The full SPN Is registered with the KDC, and the KDC provides the keytab. `gss_acquire_cred_from` finds the entry in the keytab.

BenPope commented on src/v/ssx/thread_worker.h:
```diff
@@ -0,0 +1,172 @@
+
+    typename futurator::type get_future() noexcept {
+        co_await _complete.wait();
+        co_return co_await _result.get_future();
```
Fair enough, it does look a bit weird.

BenPope replied:
Done

michael-redpanda commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+}
+
+result<void>
+gssapi_authenticator::init(ss::sstring principal, ss::sstring keytab) {
```
Should there be an assert check or something to ensure that when `init` is called we are in the `init` state?

michael-redpanda commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+    gss_key_value_element_desc elem{.key = "keytab", .value = keytab.c_str()};
+    gss_key_value_set_desc store{.count = 1, .elements = &elem};
+
+    major_status = ::gss_acquire_cred_from(
+      &minor_status,
+      server_name,
+      0,
+      GSS_C_NO_OID_SET,
+      GSS_C_ACCEPT,
+      &store,
+      &_server_creds,
+      NULL,
+      NULL);
```
having the devil of a time trying to find the documentation for this function.  I"m interested in the second to last parameter "actual_mechs".

BenPope replied:
It's not really documented, but `gss_acquire_cred` is actually just a thin wrapper over `gss_acquire_cred_from` that passes a null store. So I'm using it here to pass in the keytab rather than have it taken from an env var.

graphcareful commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+    case state::failed:
+        break;
+    }
+    vlog(seclog.info, "gss {} authenticate failed", _state);
```
`complete` follows the same code path as `failed`, is this intentional?

BenPope replied:
Yes, we don't expect to receive more auth requests after complete.

dcoded replied:
If the client attempts to reauthenticate after success what is the current behavior? Is the request to reauthenticate ignored and Redpanda returns success again, or does Redpanda then return a failure and close the connection?

BenPope replied:
If the handshake has completed, and auth is requested again, then the connection is dropped.

graphcareful commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+        return errc::invalid_credentials;
+    }
+    uint32_t ssf;
+    memcpy(&ssf, bufset[0].value(), 4);
```
nit: sizeof(uint32_t)

BenPope replied:
Done

graphcareful commented on tests/docker/Dockerfile:
```diff
@@ -162,6 +166,7 @@ COPY --chown=0:0 tests/docker/ssh /root/.ssh
 COPY --chown=0:0 tests/setup.py /root/tests/
 RUN python3 -m pip install --upgrade --force pip && \
     python3 -m pip install --force --no-cache-dir -e /root/tests/
+# RUN python3 -m pip install --no-cache-dir --no-binary confluent-kafka confluent-kafka==1.7.0 --use-pep517 -v
```
Maybe this should go in a separate commit?

BenPope replied:
Removed, it wasn't as needed as I thought it was.

graphcareful reviewed (commented):
No major comments, awesome job !

michael-redpanda commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+      &recv_tok,
+      GSS_C_NO_CHANNEL_BINDINGS,
+      &_client_name,
+      &_doid,
```
Is this correct?  I think this is for what security mechanism was used and this was populated by `gss_display_name` which is the type of the returned name.

michael-redpanda replied:
I'm referring to passing `_doid` to `gss_accept_sec_context`

BenPope replied:
I've changed this.

michael-redpanda commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+        return errc::invalid_credentials;
+    }
+
+    // OM_uint32 maj_stat = 0, min_stat = 0;
+    // gss_buffer_set_t bufset = GSS_C_NO_BUFFER_SET;
+    // gss_OID ssf_oid = ::GSS_C_SEC_CONTEXT_SASL_SSF;
+    // uint32_t ssf;
+
+    // maj_stat = gss_inquire_sec_context_by_oid(
+    //   &min_stat, _context, ssf_oid, &bufset);
```
what's our policy on commented out code?  In the past I've flag it to be deleted or putting a comment around it indicating why it's sticking around.

BenPope replied:
I'll clean this up

BenPope replied:
Done

michael-redpanda commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,514 @@
+/*
```
An overall comment, would it be possible to maybe document what the GSS calls are doing, or maybe point at the example code you used?  I was following it ok until the SSF portion.

Thanks :)

BenPope replied:
I think it was this: https://github.com/cyrusimap/cyrus-sasl/blob/master/plugins/gssapi.c

BenPope commented on src/v/ssx/thread_worker.h:
```diff
@@ -0,0 +1,172 @@
+    template<typename Func>
+    auto submit(Func func) ->
+      typename ss::futurize<std::invoke_result_t<Func>>::type {
+        return ss::smp::submit_to(
+          shard_id, [this, func{std::move(func)}]() mutable {
+              return _impl.submit(std::move(func));
+          });
```
I haven't changed this - all work is drained before stopping.

BenPope replied:
Protected with an `ss::sharded<ss::gate>`

BenPope commented on src/v/ssx/thread_worker.h:
```diff
@@ -0,0 +1,172 @@
+    }
+    ss::future<> stop() {
+        return ss::smp::submit_to(shard_id, [this]() { return _impl.stop(); });
+    }
```
Done

BenPope commented on src/v/ssx/thread_worker.h:
```diff
@@ -0,0 +1,172 @@
+        _task = std::move(task);
+        _ready.signal(RUNNABLE_READY);
+
+        co_return co_await std::move(f);
```
Done

BenPope commented on src/v/config/validators.cc:
```diff
@@ -104,4 +111,33 @@ std::optional<ss::sstring> validate_client_groups_byte_rate_quota(
+}
+
+std::optional<ss::sstring> validate_sasl_mechanisms(ss::sstring mechanisms) {
+    const absl::flat_hash_set<std::string_view> supported{"GSSAPI", "SCRAM"};
```
Done

BenPope commented on src/v/ssx/thread_worker.h:
```diff
@@ -0,0 +1,172 @@
+
+private:
+    void configure_thread() {
+        ::pthread_setname_np(::pthread_self(), "ssx::thread_worker");
```
It's a good question I don't really know the answer to.

BenPope commented on src/v/ssx/tests/thread_worker.cc:
```diff
@@ -0,0 +1,56 @@
+        }
+        BOOST_REQUIRE_EQUAL(results[i].get(), i);
+    }
+}
```
Done

BenPope commented:
Changes in force-push:
* Made sasl_mechanisms a std:vector which simplies validation
* Reduced the diff in sasl_handshake_handler::handle by not using the factories
* Make the thread_worker test multicore
* Asserted in thread_worker start & stop instead of scheduling to the shard
* Adde errc::invalid_gssapi_state
* Asynchronous sasl_mechanism::authenticate now tykes bytes instead of bytes_view
* Refactoring and cleanup in gssapi_authenticator:
  * Fixup all the error returns in gssapi_authenticator
  * Some dead-code removal
  * Logging improvements
  * Added a fail function for logging and setting state::failed
  * Simplified gssapi_authenticator::authenticate
  * Put the check and principal assignment at the end (after ssfcap/ssfreq)
  * Removed some data members
  * Reset the context and server_credential once handshake is done
* Fixes for a few nits

BenPope commented:
Changes in force-push:
* Rebase over conflict in `tests/docker/Dockerfile`

BenPope commented:
Changes in force-push:
* Workaround `cppcoreguidelines-avoid-capturing-lambda-coroutines`

BenPope commented:
Changes in force-push:
* Implement `worker_task` over alien::run_on` instead of `ss::readable_eventfd`
* Protect `thread_worker` with an `ss::sharded<ss::gate>`

BenPope commented on src/v/ssx/thread_worker.h:
```diff
@@ -0,0 +1,172 @@
+    }
+    size_t run() {
+        configure_thread();
+        while (true) {
```
Done

michael-redpanda commented on src/v/kafka/server/server.cc:
```diff
@@ -395,29 +400,59 @@ ss::future<response_ptr> sasl_handshake_handler::handle(
     vlog(klog.debug, "Received SASL_HANDSHAKE {}", request);
 
+    const auto& configured = config::shard_local_cfg().sasl_mechanisms();
+    auto supports = [&configured](std::string_view value) {
+        return absl::c_find(configured, value) != configured.end();
+    };
```
i like that :)

michael-redpanda commented on src/v/security/gssapi_authenticator.cc:
```diff
@@ -0,0 +1,422 @@
+    return outcome::success();
+}
+
+void gssapi_authenticator::finish() {
+    _context.reset();
+    _server_creds.reset();
+    _state = state::complete;
+}
```
Is there like a re-roll that's going to happen?  Or is that a TBD thing to handle?  Is that part of the 'ticket' that kerberos produces?

dotnwat commented:
### Ducktape failure triage

When `custard` is used as the scram mechanism, it looks like there is an incorrect assumption about the behavior of the admin api behavior. Probably as a result of now validating that input perhaps differently than it was before.

```
[DEBUG - 2023-01-13 22:17:08,562 - admin - _request - lineno:305]: Dispatching PUT http://docker-rp-1:9644/v1/cluster_config
[WARNING - 2023-01-13 22:17:08,564 - admin - _request - lineno:323]: Response 400: {"sasl_mechanisms":"'custard' is not a supported SASL mechanism"}
[ERROR - 2023-01-13 22:17:08,564 - cluster - wrapped - lineno:41]: Test failed, doing failure checks...
Traceback (most recent call last):
  File "/root/tests/rptest/services/cluster.py", line 35, in wrapped
    r = f(self, *args, **kwargs)
  File "/root/tests/rptest/tests/cluster_config_test.py", line 581, in test_valid_settings
    patch_result = self.admin.patch_cluster_config(upsert=updates,
  File "/root/tests/rptest/services/admin.py", line 376, in patch_cluster_config
    return self._request("PUT",
  File "/root/tests/rptest/services/admin.py", line 336, in _request
    r.raise_for_status()
  File "/usr/local/lib/python3.10/dist-packages/requests/models.py", line 941, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
```

dotnwat commented:
* Failure is k8s

dotnwat commented:
/ci-repeat 10 skip-units

dotnwat commented:
/ci-repeat 10 skip-units

dotnwat commented:
* Failure https://github.com/redpanda-data/redpanda/issues/8218

dotnwat reviewed (commented):
looks awesome. i think there are a couple minor issues with worker thread shutdown, but that's about it.

BenPope commented on src/v/config/configuration.cc:
```diff
@@ -829,6 +829,25 @@ configuration::configuration()
+      "The location of the Kerberos keytab file for Redpanda",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      "/var/lib/redpanda/redpanda.keytab")
+  , sasl_kerberos_principal(
```
Since per-node config is not desirable, I've made it work this way:

The SPN is made if 3 parts, `<primary>/<host>@<realm>`.

The primary should be specified here.

We're using `gss_import_name` with `GSS_C_NT_HOSTBASED_SERVICE`. 

So if the KDC has an entry for each host, and that entry is requested for the keytab, then no node-level configuration is required in Redpanda, it should be found during `gss_acquire_cred_from`.

This is how the test does it.

dotnwat replied:
:+1:

BenPope commented on src/v/ssx/thread_worker.h:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    void process(ss::alien::instance& alien, ss::shard_id shard) final {
+        auto f = futurator::invoke(_func);
+        ss::alien::run_on(alien, shard, [this, f{std::move(f)}]() mutable {
+            f.forward_to(std::move(_promise));
```
It should work like this:

1. On reactor shard 0:
  a. Build the `worker_task` hat contains the promise.
  b. Run `worker_task::get_future()`.
2. On `std::thread`
  a. Run `worker_task::process` (`futurize::invoke` is synchronous)
3. On reactor shard 0:
  a. Satisfy `worker_task::_promise` with the result.

The `_promise` and its `future` are only accessed by shard 0.

Perhaps the `future` from `futurize::invoke` is confusing here? It's being used only as an intermediate value to carry the ready "result-or-exception" between threads.

dotnwat replied:
let's run _func without futurator::invoke. i think that it is safe in its current form, but visually it should give someone pause and go read all of futurator::invoke to make sure it is safe to call from a non-reactor thread. given that we just need to run func and capture its value or an exception and move it into the continuation passed to alien::run_on, i'd vote for keeping the alien thread free of all future/promise bits.

BenPope commented on src/v/ssx/thread_worker.h:
```diff
@@ -0,0 +1,224 @@
+        co_await _gate.close();
+        _ready.signal(RUNNABLE_STOP);
+        if (_worker.joinable()) {
+            _worker.join();
```
I guess I could be more explicit here - it's certainly designed for start and stop to be tied to the application lifetime.

The idea is to async wait for all operations to finish with `gate::wait`, so there shouldn't be any significant work left on the thread at the time it's joined, just destruction.

dotnwat replied:
>significant work left on the thread at the time it's joined, just destruction.

join waits for the thread to exit, so i'd assume it might block and cause a reactor stall. i don't know if that is true, and that's not a problem since it's tied to app lifetime, but something i would expect we write a comment about to warn users.

BenPope commented:
Changes in force-push:
* Add a message for `errc::invalid_gssapi_state` (and remove the default case so compiler catches missing ones)
* Improve shutdown code for `impl::thread_worker::run` the task should never have a value, but it's clearer now.
* Wait on the gate in `thread_worker::stop()`
